### PR TITLE
RavenDB-26449 RavenDB Analyzers

### PR DESCRIPTION
**Pull request by Scooletz**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-26449/RavenDB.Analyzers

### Additional description

This PR introduces a set of Roslyn diagnostic analyzers (and companion code fixes) that run at compile-time inside any project that references `RavenDB.Client`. They catch common misuse patterns early even without tests and before running the actual application

#### New Diagnostics

Each rule declares in code the version it was introduced in and the severity it should eventually reach. In the release that introduces a rule, it ships as `DiagnosticSeverity.Info` so that **it does not break existing builds**. The rule is automatically promoted to its destination severity in the next release. Users can still override any rule in their own `.editorconfig` or `Directory.Build.props`.

The **Fix** column marks rules with an automated code fix accessible via the IDE lightbulb / `dotnet fix`. The remaining rules do not ship fixes because the correct resolution depends on application intent or they are hard to fix automatically.

| ID | Title | What it catches | Fix | Manual Verification |
|----|-------|-----------------|:---:| --- |
| **RVN001** | Index Map/Reduce assigned outside constructor | `Map =` or `Reduce =` written in a method rather than the constructor; the expression may never be registered | — | Yes. Using `Library` sample |
| **RVN002** | Query operator after projection | `Where`, `OrderBy`, `Search`, etc. placed after `.ProjectInto` or `.Select`; the operator silently targets the projected shape instead of the source document | — | Yes. Using `Library` sample |
| **RVN003** | Double ProjectInto | `.ProjectInto` called twice on the same chain; throws `InvalidOperationException` at runtime | — | Yes. Using `Library` sample  |
| **RVN004** | Missing Map assignment | `AbstractIndexCreationTask<T>` subclass with no `Map =` in any constructor; index deployed with no definition | — | Yes. Using `Library` sample  |
| **RVN005** | Multi-map missing AddMap | `AbstractMultiMapIndexCreationTask` subclass with no `AddMap` call; same as above | — | Yes. Using `Library` sample  |
| **RVN006** | Multi-map with single AddMap | Only one `AddMap` call present; a plain `AbstractIndexCreationTask<T>` would be simpler | — | Yes. Using `Library` sample  |
| **RVN007** | Query field not indexed | `Where`/`OrderBy`/`Search` lambda references a field not in the index projection; query returns no results for that field | — |
| **RVN008** | Projected field not retrievable | Projected field is absent from both the stored index entry and the source document; value will be `null` at runtime | — |
| **RVN009** | Unsupported method in index Map/Reduce | User-defined helper method called inside a `Map`/`Reduce` lambda; cannot be translated to server-side IL | — |
| **RVN010** | Unsupported method in query expression | User-defined method called inside a `Where`/`OrderBy`/`Select` lambda; throws at runtime during RQL translation | — |
| **RVN011** | Wrong session source in subscription | `store.OpenSession()` / `store.OpenAsyncSession()` called inside a subscription `Run` lambda; must use `batch.OpenSession()` instead to participate in the batch acknowledge transaction | ✅ rewrites `store.OpenSession()` → `batch.OpenSession()` (and async) |
| **RVN012** | Session operations should be batched (lazy API) | Two or more independent `Load` / `Query` calls in the same method; can be combined into a single multi-get request using the lazy API | ✅ converts the eager operations to lazy registrations (`Lazily.Load`/`LoadAsync`, `Lazily()`/`LazilyAsync()`) and reads each `.Value` — realizing the first value dispatches the whole batch in one round-trip, so no explicit `ExecuteAllPendingLazyOperations` call is emitted; bails out entirely if any statement can't be safely rewritten |
| **RVN013** | Unbounded query result | Materializing query call (`ToList`, `ToArray`, async variants) with no `.Take(n)`; intent is implicit and may fetch far more data than intended as the dataset grows | — | Yes. Using `Library` sample | 
| **RVN014** | Index Map fan-out | `SelectMany` or nested `from` inside a `Map`/`AddMap` lambda; each source document yields one entry per collection element — mirrors the existing server-side `WarnIndexOutputsPerDocument` warning | — |

Every rule carries a help link of the form `https://ravendb.net/l/<code>` that resolves to its documentation page; hovering a squiggle in the IDE exposes the link directly. Using a redirector lets the doc target move without shipping a new analyzer.

#### Index metadata source generator

An index declares its shape in a constructor body, and constructor bodies are not part of compiled metadata, so analyzers could not inspect an index declared in another project: silently nothing on the command line, and `AD0001` in the IDE, where a `CompilationReference` does carry syntax but belongs to another compilation. Attributes *are* metadata, so `IndexMetadataGenerator` (an `IIncrementalGenerator` in `Raven.Analyzers.dll`) runs in the project declaring the index and records its shape as `[assembly: RavenIndexMetadata(typeof(BooksByAuthor), MapFields = ..., AssignsMap = ..., ...)]`, covering the whole inheritance chain and including abstract bases. RVN004, RVN005, RVN006, RVN007, RVN008 and RVN009 now read that attribute instead of index source, so they behave identically for a local index and a referenced DLL; `Analyzable = false` (JavaScript index, `CreateField`, block-bodied `Map`) means *unknown*, so nothing is reported. `RavenIndexMetadataAttribute` is new public API on `Raven.Client`, marked `[EditorBrowsable(Never)]`.

#### Distribution

Both analyzer assemblies (`Raven.Analyzers.dll` and `Raven.Analyzers.CodeFixes.dll`) are embedded in `RavenDB.Client.nupkg` under `analyzers/dotnet/cs/`. They ship automatically with every client package update, no separate install required.

The analyzers and code fixes are kept compiler-safe (RS1038: they reference only compiler-provided assemblies), so they load cleanly during command-line builds. The code-fix assembly's `Workspaces` dependency is only exercised in the IDE/`dotnet fix` host.

Both assemblies pin `Microsoft.CodeAnalysis` to **4.4.0**, deliberately below the repo-wide version: an analyzer built against a newer Roslyn than the host compiler fails to load with `CS9057`, which is a build *failure* in the consumer's project under `TreatWarningsAsErrors`. That pin also requires transitive package pinning to be off in both projects (or the analyzer fails to load with `CS8032`) and `ImmutableArray.Create(...)` instead of collection expressions; see `src/Raven.Analyzers/readme.md`. Verified on the .NET 8, 9 and 10 SDKs with `TreatWarningsAsErrors=true`.

#### Severity Philosophy

New rules ship at `Info`, then auto-promote to their destination severity in the next release automatically.

Each diagnostic declares two things at its definition site (`DiagnosticDescriptors.Create`): the version it was **introduced at** (`introducedAt`, e.g. `"7.2"`) and its **destination severity** (`Warning` for actionable correctness rules; `Info` for the heuristic/style advisories RVN006, RVN007, RVN008). The effective `DefaultSeverity` is then computed by `SeverityPolicy` against the product version baked into the analyzer assembly at build time (the `GenerateRavenProductVersion` target reads the canonical `AssemblyVersion` from `src/CommonAssemblyInfo.cs`):

When a rule is introduced in a release it defaults to `Info`. Existing codebases see hints but do not break on upgrade, and teams get a window to address occurrences and report false positives. 

Once the version moves past that release the rule is automatically its destination severity. The promotion cadence follows the precision of `introducedAt`: a minor version like `"7.2"` is compared at major.minor granularity and promotes at `7.3.0`; a patch version like `"7.2.5"` is compared at major.minor.patch granularity and promotes at `7.2.6`.

All 14 rules in this PR are introduced at `7.2`, so on the current `7.2.x` builds every rule is `Info`. On `7.3.0` the 11 correctness rules become `Warning` automatically while a few of them stay `Info`. This happens automatically, during the build process.

#### Testing

`test/AnalyzersTests` contains Roslyn unit tests that exercise every rule and every code fix, including positive cases, negative cases, async variants, and edge cases. `DiagnosticSeverityPolicyTests` additionally enforces the version-aware severity policy described above.

`CrossProjectIndexRegressionTests` covers indexes declared in a different project, running each case through both a compiled DLL and a `CompilationReference` and failing on `AD0001` rather than letting a crashed analyzer read as "no diagnostics". `IndexMetadataGeneratorIncrementalityTests` asserts the generator's pipeline steps are reused across an unrelated edit, which nothing else would catch. Tests use `[RavenFact]` / `[RavenTheory]` with `RavenTestCategory.ClientApi` but do not inherit `RavenTestBase`, since they need no server; the project is not referenced from `SlowTests`, so `TestsInheritanceTests` never scans it.

#### Operational Notes

- `test/AnalyzersTests` is fast (a few seconds) and should be added as a CI preflight gate before
  `FastTests`.
- `AnalyzersTests.csproj` carries a comment explaining why the project uses `[RavenFact]` without
  inheriting `RavenTestBase`, so future maintainers do not wire it into the server test base classes.
- Bumping `Microsoft.CodeAnalysis` in `Directory.Packages.props` does **not** move the analyzers.
  Their pin is deliberate and raising it drops consumer SDKs; see the readme before changing it.
- New analyzer rules should be documented in release notes listing the rule ID, title, and
  default severity.

#### Benchmark

The benchmark project lives at `bench/Raven.Analyzers` (alongside the other `*.Benchmark` projects). It compiles a body of realistic indexes and queries so the analyzers have real work to do. Its `compare.ps1` measures the build overhead the analyzers add.

`compare.ps1` interleaves several clean builds per case (5 by default: without, with, without, with, ...) so a transient IO or CPU hiccup hits both cases rather than skewing one, and reports the median of each. It writes `analyzers-benchmark-result.json` next to the script (git-ignored):

```json
{
  "without": 2.847,
  "with": 4.213,
  "overheadAbsolute": 1.366,
  "overheadRelative": 47.98,
  "success": false
}
```

`without` and `with` are the median wall-clock seconds, `overheadAbsolute` is their difference in seconds, `overheadRelative` is that difference as a percentage, and `success` is the gate: `true` when the relative overhead stays under 20%.

image

And the actual reports

image

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

**Comment by JamesFieldist** [comment]
Unfortunately when I click on the link to the issue it goes nowhere for me (404).  Is there any information on what analysers this adds?

**Comment by Scooletz** [comment]
> Unfortunately when I click on the link to the issue it goes nowhere for me (404). Is there any information on what analysers this adds?

@JamesFieldist  I'm collecting them in a readme under https://github.com/Scooletz/ravendb/blob/raven-analyzers/src/Raven.Analyzers/Readme.md for now. Once I finalize the initial scope, I'll update the PR etc.

**Comment by Scooletz** [comment]
I've updated the description so that it shows **what** diagnostics and fixes and **why** they are introduced. I want to ship soon. Fixing final remarks and issues.

/FYI @JamesFieldist

**Comment by Scooletz** [comment]
Labeled as `area-ClientApi` to not burn the CI/CD

**Comment by JamesFieldist** [comment]
Question for you: My understanding is that if you embed analysers only the projects that directly refer to the nuget that embeds them will have the analysers execute.

Often you have other projects that reference that library, and you want to extend the analysers. So most projects like mass transit ALSO have a separate nuget package that you can reference directly in every project that doesn't directly reference the root nuget but you still want to activate the analysers.

Thoughts?

**Comment by JamesFieldist** [comment]
Also on RVN010:

I've never figured out how to reliably create reusable mappers that are Expression<Func...>> Sometimes they work, sometimes they don't. It would be nice if the docs were updated and this robustly identified these cases.

**Comment by JamesFieldist** [comment]
And other feedback:

If you use anything other than new Array[] {} for an empty array, in my experience the index will fail.

You can't use new notation of [], or even new Array[0]. They both throw errors in the index for me.  Ideally both of these notations would be supported by the index creation, but until then it would be really nice for these to be an error in the analyser since the indexes won't work.

Same for ??, ??=. Neither work. It would be really nice if they did, but until they do, they should throw an error in the analyser.

**Comment by JamesFieldist** [comment]
And I think this is the last one, and it may be covered by RVN008

We should get an error if the projection type can't be mapped through the inheritence chain. I.e. if the result type is stored and it inherits from a parent type, you can project into the result type, the parent type or the document. You can't project into anything else (unless, by fluke the type as the same fields but has no inheritence)

It was unclear in the readme if it is enforcing those.  The later case would be pretty expensive I would think to validate, but boy would it be nice.

**Comment by Scooletz** [comment]
> Question for you: My understanding is that if you embed analysers only the projects that directly refer to the nuget that embeds them will have the analysers execute.
> 
> Often you have other projects that reference that library, and you want to extend the analysers. So most projects like mass transit ALSO have a separate nuget package that you can reference directly in every project that doesn't directly reference the root nuget but you still want to activate the analysers.
> 
> Thoughts?

The plan is to embed them in `RavenDB.Client` itself. The adoption of on a non-embedded analyzers looks much much worse. We want people to use them.

**Comment by Scooletz** [comment]
> Also on RVN010:
> 
> I've never figured out how to reliably create reusable mappers that are Expression<Func...>> Sometimes they work, sometimes they don't. It would be nice if the docs were updated and this robustly identified these cases.

Do you mean updating the docs **as well** or something else?

**Comment by Scooletz** [comment]
> And other feedback:
> 
> If you use anything other than new Array[] {} for an empty array, in my experience the index will fail.
> 
> You can't use new notation of [], or even new Array[0]. They both throw errors in the index for me. Ideally both of these notations would be supported by the index creation, but until then it would be really nice for these to be an error in the analyser since the indexes won't work.
> 
> Same for ??, ??=. Neither work. It would be really nice if they did, but until they do, they should throw an error in the analyser.

I see. Not sure if I can tackle it. I'll see.

**Comment by Scooletz** [comment]
> And I think this is the last one, and it may be covered by RVN008
> 
> We should get an error if the projection type can't be mapped through the inheritence chain. I.e. if the result type is stored and it inherits from a parent type, you can project into the result type, the parent type or the document. You can't project into anything else (unless, by fluke the type as the same fields but has no inheritence)
> 
> It was unclear in the readme if it is enforcing those. The later case would be pretty expensive I would think to validate, but boy would it be nice.

I think I need to build up an actual test for the inheritance chain.

**Comment by JamesFieldist** [comment]
> > Also on RVN010:
> > I've never figured out how to reliably create reusable mappers that are Expression<Func...>> Sometimes they work, sometimes they don't. It would be nice if the docs were updated and this robustly identified these cases.
> 
> Do you mean updating the docs **as well** or something else?

It would be nice if the docs for ravendb tackled this. Mappers are really common and RavenDb is really flaky with them.  Even manually creating your mappers (which you should) is unreliable in a lot of cases from my experience (especially if they have RavenQuery in them, which probably is related to indexes don't allow you to project with RavenQuery from the base document).  An explainer on how to use them in the docs would be great, and the analyser warning when your shared Expression<Func is going to break would be ideal if possible.

**Comment by Scooletz** [comment]
> > > Also on RVN010:
> > > I've never figured out how to reliably create reusable mappers that are Expression<Func...>> Sometimes they work, sometimes they don't. It would be nice if the docs were updated and this robustly identified these cases.
> > 
> > 
> > Do you mean updating the docs **as well** or something else?
> 
> It would be nice if the docs for ravendb tackled this. Mappers are really common and RavenDb is really flaky with them. Even manually creating your mappers (which you should) is unreliable in a lot of cases from my experience (especially if they have RavenQuery in them, which probably is related to indexes don't allow you to project with RavenQuery from the base document). An explainer on how to use them in the docs would be great, and the analyser warning when your shared Expression<Func is going to break would be ideal if possible.

Thank you. Will be moving forward with it soon.

File: src/Raven.Analyzers/Indexes/IndexDefinitionAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        private static void CheckForMapReduceAssignmentsInMethod(
+            SyntaxNodeAnalysisContext context,
+            MethodDeclarationSyntax method)
+        {
+            if (method.Body == null)
+                return;
+
+            foreach (AssignmentExpressionSyntax assignment in
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/Indexes/IndexDefinitionAnalyzer.cs:128]: Addressed. CheckForMapReduceAssignmentsInMethod now uses method.GetBodyNode(), which also covers expression-bodied members (=>), so Map/Reduce assignments outside the constructor are flagged there too.

File: test/AnalyzersTests/AnalyzersTests.csproj [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+    <AssemblyName>AnalyzersTests</AssemblyName>
+    <OutputType>Exe</OutputType>
+    <Nullable>enable</Nullable>
+    <LangVersion>preview</LangVersion>
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply test/AnalyzersTests/AnalyzersTests.csproj:8]: Addressed. AnalyzersTests.csproj now sets RuntimeFrameworkVersion and CodeAnalysisRuleSet (..\..\RavenDB.ruleset). Pinned to 10.0.9 to match this project's net10 / xUnit v3 toolchain rather than 10.0.6.

File: src/Raven.Analyzers/KnownTypes.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        [
+            "ToList", "ToListAsync", "ToArray", "ToArrayAsync",
+        ];
+
+        // Index class names (matched by short name against base type chain)
+        public const string AbstractIndexCreationTaskGenericName = "AbstractIndexCreationTask";
+        public const string AbstractGenericIndexCreationTaskName = "AbstractGenericIndexCreationTask";
+        public const string AbstractMultiMapIndexCreationTaskName = "AbstractMultiMapIndexCreationTask";
+        public const string AbstractMultiMapTimeSeriesIndexCreationTaskName = "AbstractMultiMapTimeSeriesIndexCreationTask";
+        public const string AbstractMultiMapCountersIndexCreationTaskName = "AbstractMultiMapCountersIndexCreationTask";
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/KnownTypes.cs:120]: By design for now. IsIndexCreationTask covers document and multi-map bases. Regular AbstractTimeSeriesIndexCreationTask / AbstractCountersIndexCreationTask are intentionally not matched yet because their map conventions differ and broadening would risk false positives. The XML doc was corrected to state the actual scope; broadening is a follow-up.

File: src/Raven.Analyzers/Shared/SyntaxHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        {
+            INamedTypeSymbol? current = symbol?.BaseType;
+            while (current != null)
+            {
+                string name = current.Name;
+                if (name is KnownTypes.AbstractIndexCreationTaskGenericName
+                         or KnownTypes.AbstractGenericIndexCreationTaskName
+                         or KnownTypes.AbstractMultiMapIndexCreationTaskName
+                         or KnownTypes.AbstractMultiMapTimeSeriesIndexCreationTaskName
+                         or KnownTypes.AbstractMultiMapCountersIndexCreationTaskName)
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/Shared/SyntaxHelpers.cs]: Addressed by correcting the XML doc to match the actual scope (document and multi-map bases). Regular time-series/counters bases are intentionally out of scope for now to avoid false positives; broadening is tracked as a follow-up.

File: src/Raven.Analyzers/Subscriptions/SubscriptionOpenSessionAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            {
+                // A non-static local function captures the enclosing Run lambda's batch
+                // parameter, so an OpenSession inside it is still convertible to
+                // batch.OpenSession(): treat it as transparent and keep walking up. A static
+                // local function cannot capture batch, so it is a real scope boundary.
+                if (current is LocalFunctionStatementSyntax localFunction)
+                {
+                    if (localFunction.Modifiers.Any(SyntaxKind.StaticKeyword))
+                        return false;
+                }
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/Subscriptions/SubscriptionOpenSessionAnalyzer.cs:79]: We are keeping the current behavior here by design. A `static` local function is treated as a scope boundary, so RVN011 is not reported inside it. A static local function cannot capture the Run lambda's `batch` parameter and is a reusable, context-free helper that may be called from outside the subscription flow, so flagging it purely on lexical nesting would be a false positive (and the suggested fix would not even compile, since `batch` is out of scope). The code fix already bails for static local functions. Confirmed with the maintainer.

[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/Subscriptions/SubscriptionOpenSessionAnalyzer.cs:79]: We are keeping the current behavior here by design. A `static` local function is treated as a scope boundary, so RVN011 is not reported inside it. A static local function cannot capture the Run lambda's `batch` parameter and is a reusable, context-free helper that may be called from outside the subscription flow, so flagging it purely on lexical nesting would be a false positive (and the suggested fix would not even compile, since `batch` is out of scope). The code fix already bails for static local functions. Confirmed with the maintainer.

File: src/Raven.Analyzers/Subscriptions/SubscriptionOpenSessionAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Raven.Analyzers.Shared;
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/Subscriptions/SubscriptionOpenSessionAnalyzer.cs:6]: This is a false positive. `localFunction.Modifiers` is a `SyntaxTokenList`, and `SyntaxTokenList.Any(SyntaxKind)` is a built-in instance overload from Microsoft.CodeAnalysis, not `Enumerable.Any`, so `System.Linq` is not required. The file does not import System.Linq and the analyzer compiles cleanly with warnings-as-errors; adding the import would instead introduce an unused-using (CS8019).

[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/Subscriptions/SubscriptionOpenSessionAnalyzer.cs:6]: Modifiers.Any(SyntaxKind) is the built-in SyntaxTokenList overload, not LINQ

File: src/Raven.Analyzers/Queries/QueryProjectionFieldAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+                    return "bail"; // variable or computed — bail
+
+                string typeIdent = behaviorAccess.Expression switch
+                {
+                    IdentifierNameSyntax id => id.Identifier.Text,
+                    _ => string.Empty
+                };
+
+                if (typeIdent != KnownTypes.ProjectionBehaviorTypeName)
+                    return "bail";
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Raven.Analyzers/Queries/QueryProjectionFieldAnalyzer.cs]: Not gonna do it
